### PR TITLE
feat(ports): add ports command to list and check host port bindings

### LINKED ISSUE #509: Add azd app ports command to list host ports and flag conflicts
## Problem

Finding which port a service listens on means starting the app and reading the dashboard or logs. There is no static view of the host port each service is configured to use, and no warning when two services ask for the same host port.

## Solution

Add `azd app ports` that reads azure.yaml and prints each service with its configured host port. Services without an explicit port show as auto-assigned. Report any host port claimed by more than one service, and exit non-zero when a duplicate explicit port is found so it can gate a script. Support `--output json`.

## Acceptance criteria

- Prints each service with its explicit host port, or "auto" when none is set
- Detects and reports duplicate explicit host ports across services
- Exits non-zero when a duplicate explicit host port is found
- `--output json` emits a machine-readable list
- Unit tests cover explicit ports, auto ports, and the duplicate-port case

### PR DESCRIPTION
**Pull request by jongio**
## What

Adds `azd app ports`, a read-only command that lists the host port each service binds, resolved from azure.yaml.

For every service it prints each binding as `host -> container/protocol`. An explicit host port shows as its number; a port left for the runtime to assign shows as `auto`. When two bindings claim the same explicit host port, the command marks the conflict, prints a warning, and exits non-zero. That makes it a quick preflight check before `azd app run` and lets it gate a pipeline in CI.

## Why

There is no fast way to see, across all services, which host ports are claimed and whether two services collide on the same one until `run` fails to bind. Listing the resolved ports up front, and failing on duplicate explicit bindings, catches the collision before startup.

## Usage

```bash
# List host ports for every service
azd app ports

# JSON object keyed by service name
azd app ports --output json
```

Text output:

```
web
  port: 3000 -> 8080/tcp  (conflict)
worker
  port: 3000 -> 9090/tcp  (conflict)

Host port 3000 is bound by more than one service: web, worker
```

## Details

- Reuses the existing port model (`GetPortMappings`, `ParsePortSpec`) so protocol suffixes, `ip:host:container` forms, and Docker auto-assign all resolve the same way `run` sees them.
- Only explicit host ports can conflict; auto-assigned ports never count, since the runtime picks a free port for each.
- Conflicts are detected across services and within a single service, deduped by owner, and sorted by port for stable output.
- JSON mode prints an object keyed by service name for programmatic use.

## Tests

Table-driven unit tests cover the basic report, cross-service conflict, auto ports never conflicting, intra-service duplicate, UDP protocol, JSON shape, and the conflict summary and helper functions.

## Verification

- `go build ./...` clean
- `go vet ./src/cmd/app/commands/...` clean
- `go test ./src/cmd/app/commands/` clean
- `golangci-lint run ./src/cmd/app/commands/...` reports 0 issues

Closes #509

### CODE REVIEW

File: cli/src/cmd/app/commands/ports.go [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+		bindings := make([]portBinding, 0, len(mappings))
+		for _, m := range mappings {
+			b := portBinding{
+				Container: m.ContainerPort,
+				Protocol:  protocolOrDefault(m.Protocol),
+			}
+			if m.HostPort > 0 {
+				b.HostPort = m.HostPort
+				b.Host = strconv.Itoa(m.HostPort)
+				owners[m.HostPort] = append(owners[m.HostPort], name)
```
Comment (wbreza) [inline cli/src/cmd/app/commands/ports.go]: **[conflict-detection] (High)** Conflicts are keyed only on the numeric host port (`owners map[int][]string`), ignoring `PortMapping.BindIP` and `Protocol`. `ParsePortSpec` populates both, so valid, non-colliding configurations are flagged and cause a non-zero exit:

- Different bind IPs on the same port — e.g. `127.0.0.1:3000:8080` and `192.168.1.5:3000:8080` — do not collide at the OS level but are reported as a conflict.
- Same port, different protocol — e.g. `3000:8080/tcp` and `3000:9090/udp` — can coexist but are reported as a conflict.

Because this command is meant to gate `azd app run` and CI on real conflicts (non-zero exit), these false positives fail valid setups. Key ownership on the full binding identity `(BindIP, HostPort, Protocol)`: treat wildcard binds (`0.0.0.0` / `::` / empty) as overlapping any specific IP on the same port+protocol, and treat two distinct specific IPs as non-conflicting. Apply the same identity when setting the per-binding `Conflict` flag and building the summary.

File: cli/src/cmd/app/commands/ports.go [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+		Args:         cobra.NoArgs,
+		RunE:         runPorts,
+	}
+
+	return cmd
+}
+
+// portBinding describes one host-to-container port binding for a service.
+type portBinding struct {
+	Host      string `json:"host"`               // explicit host port number, or "auto"
```
Comment (wbreza) [inline cli/src/cmd/app/commands/ports.go:46]: **[output-bindip] (Medium)** `PortMapping.BindIP` is parsed upstream but dropped here: `portBinding` has no bind-IP field, the text line renders only `host -> container/protocol`, and the JSON payload has no `bindIP`. A `127.0.0.1:8080:80` binding is indistinguishable from `8080:80` in the output, which hides the actual binding surface this command exists to surface.

Add `BindIP string ` + '`' + 'json:"bindIP,omitempty"' + '`' + `, populate it from `m.BindIP`, and render the text form as `[bindIP:]host -> container/protocol` (e.g. `127.0.0.1:8080 -> 80/tcp`).

File: cli/src/cmd/app/commands/ports.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+
+	azureYaml, err := service.ParseAzureYaml(cwd)
+	if err != nil {
+		return fmt.Errorf("failed to load azure.yaml: %w", err)
+	}
+
+	report := collectPortReport(azureYaml)
+
+	if cliout.IsJSON() {
+		if err := cliout.PrintJSON(report.services); err != nil {
```
Comment (wbreza) [inline cli/src/cmd/app/commands/ports.go]: **[json-error-channel] (Low)** In JSON mode the command writes valid JSON to stdout and then `runPorts` returns an error (lines 91-92) that cobra renders as `Error: ...` on stderr. The JSON payload also omits any top-level `conflicts` list, so JSON consumers must parse stderr to learn a conflict occurred, and scripts that merge stderr into stdout get corrupted JSON.

Include a top-level `conflicts` array in the emitted JSON and rely on the non-zero exit code to signal failure, so stdout stays valid JSON and consumers don't depend on stderr text.

**Review by wbreza — COMMENTED** [review]
## Code Review — `azd app ports`

Solid, well-scoped addition: the command cleanly reuses the existing `GetPortMappings` / `ParsePortSpec` port model and `sortedServiceNames`, produces deterministic output, and ships with table-driven tests and full docs.

One **High** issue drives this as a Comment: conflict detection keys only on the numeric host port, so valid configurations that differ by bind IP or protocol are reported as conflicts and cause a non-zero exit — which matters because the command is meant to gate `azd app run` and CI. Two independent review models flagged this. A **Medium** (bind IP dropped from output) and a **Low** (JSON mode error channel) follow.

### Summary
| Priority | Count |
|----------|-------|
| High | 1 |
| Medium | 1 |
| Low | 1 |
| **Total** | **3** |

### What looks good
- Reuses the established port model so protocol suffixes, `ip:host:container`, and Docker auto-assign resolve exactly as `run` sees them.
- Conflicts are sorted and owner-deduped for stable output; tests cover cross-service, intra-service, auto-never-conflict, UDP, and JSON shape.
- Docs (`commands/ports.md` + `cli-reference.md`) are thorough.

Details are in the inline comments.

File: cli/src/cmd/app/commands/ports.go [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+	if a.HostPort != b.HostPort || a.Protocol != b.Protocol {
+		return false
+	}
+	return bindIPsOverlap(a.BindIP, b.BindIP)
+}
+
+func bindIPsOverlap(a, b string) bool {
+	a = strings.TrimSpace(a)
+	b = strings.TrimSpace(b)
+	if a == b {
```
Comment (wbreza) [inline cli/src/cmd/app/commands/ports.go:271]: **[ipv6-textual-equivalence] (Medium)** For two specific (non-empty, non-wildcard) bind IPs, `bindIPsOverlap` decides overlap by raw string equality (`if a == b`) and otherwise returns `false`. Because `ParsePortSpec` stores the bind IP as the raw, un-normalized text from the spec, the same IPv6 address written in two textual forms is treated as non-overlapping — a deterministic false negative for a real OS-level collision:

- `[2001:db8::1]:3000:8080` vs `[2001:0db8:0:0:0:0:0:1]:3000:8080` → same address, `a == b` is false, neither side is empty/wildcard → returns `false`, so the conflict is missed.
- The same applies to non-canonical IPv4 forms (e.g. `127.0.0.1` vs `127.000.000.001`).

This undercuts the command's stated purpose as a preflight gate for exactly the bind-IP comparison this commit set out to make correct. Fail-safe direction (it under-reports rather than blocking a valid config), hence Medium not High.

Fix: compare concrete IP literals by canonical value, not string. Parse each non-wildcard side with `net.ParseIP` and compare the resulting `net.IP` (e.g. via `ip1.Equal(ip2)`) before falling through to `return false`; keep the existing wildcard/empty handling.

**Review by wbreza — COMMENTED** [review]
## Re-review — `azd app ports` (updated: `fix(ports): key conflicts on bind IP and protocol`)

The update cleanly resolves all three findings from the prior review, confirmed by both review models:

- **[Resolved · was High]** Conflict detection now keys on `(BindIP, HostPort, Protocol)` via `portBindingKey` and a union-find overlap pass. Distinct specific IPs no longer collide, and `tcp` vs `udp` on the same port no longer collides. IPv4/IPv6 wildcard-vs-specific semantics are handled explicitly and covered by the new `TestCollectPortReportConflictIdentityMatrix`.
- **[Resolved · was Medium]** `BindIP` is now carried on `portBinding` and `portConflict` and rendered in both text (`formatHostBinding`, bracketed for IPv6) and JSON. Covered by `TestPortReportTextIncludesBindIP`.
- **[Resolved · was Low]** JSON mode prints the report (now `{services, conflicts}` with a nil-normalized `conflicts: []`) then returns the error with `cmd.SilenceErrors = true`, so stdout stays valid JSON and the non-zero exit is preserved. Covered by `TestRunPortsJSONConflictOutputIsValidJSON`.

One new **Medium** correctness gap remains in the specific-IP comparison path (inline). It fails safe (under-reports a conflict rather than blocking a valid config) and needs atypical input, but it lives in the same comparison logic this commit set out to make correct, so it is worth closing.

Per this autopilot policy, Request Changes is not used; publishing as a Comment.

### Summary
| Priority | Count |
|----------|-------|
| Medium | 1 |
| **Total** | **1** |

### What looks good
- The union-find grouping is sound: transitive grouping only happens through `portBindingsOverlap`, which requires equal `HostPort` and `Protocol`, so reporting one `HostPort`/`Protocol` per group is correct.
- Thorough identity-matrix tests plus text-BindIP and valid-JSON-on-conflict coverage.
- Docs (`commands/ports.md`) updated to match the new `[bindIP:]host` format, IP-family conflict rules, and `{services, conflicts}` JSON shape.

_Note: automated build/vet/test were not re-run in place because the local working tree was not on the PR head; the review is based on direct inspection of the PR-head source and tests, and the author reports `go build`, `go vet`, `go test`, and `golangci-lint` all clean._

File: cli/src/cmd/app/commands/ports.go (hunk 1 above)
Comment (jongio) [reply cli/src/cmd/app/commands/ports.go]: Fixed in [44986204](https://github.com/jongio/azd-app/commit/44986204). Conflicts are now keyed on `portBindingKey{BindIP, HostPort, Protocol}` instead of the host port alone, so `127.0.0.1:8080` and `192.168.1.5:8080` no longer report a false conflict, and a TCP binding no longer collides with a UDP one on the same port. Overlap against the wildcard binds is handled explicitly per IP family.

File: cli/src/cmd/app/commands/ports.go (hunk 2 above)
Comment (jongio) [reply cli/src/cmd/app/commands/ports.go:46]: Fixed. `portBinding` now carries `BindIP string` with a `json:"bindIP,omitempty"` tag populated from `m.BindIP`, and the text form renders through `formatHostBinding`, which prefixes the bind IP when present and brackets IPv6 literals (`[::1]:8080`). `127.0.0.1:8080:80` is now distinguishable from `8080:80` in both output modes.

File: cli/src/cmd/app/commands/ports.go (hunk 3 above)
Comment (jongio) [reply cli/src/cmd/app/commands/ports.go]: Fixed. `portJSONReport` now carries a top-level `conflicts` array, and JSON mode sets `cmd.SilenceErrors` so cobra doesn't also print the error text. Stdout stays valid JSON, and the non-zero exit code still signals the conflict, so scripts that merge stderr into stdout no longer get corrupted output.

File: cli/src/cmd/app/commands/ports.go (hunk 4 above)
Comment (jongio) [reply cli/src/cmd/app/commands/ports.go:271]: Fixed in [1aa2fd92](https://github.com/jongio/azd-app/commit/1aa2fd92). `bindIPsOverlap` now parses both sides with `net.ParseIP` and compares with `net.IP.Equal`, so `::1` and `0:0:0:0:0:0:0:1` match, as do IPv4-mapped forms like `::ffff:127.0.0.1` and `127.0.0.1`. That mapped case is a real OS-level collision, so treating it as equal is correct. Unparseable hostnames still return false, preserving the previous behavior. Added tests for equivalent IPv6, IPv4-mapped IPv6, and distinct IPv6.

**Review by wbreza — APPROVED** [review]
## Re-review — `azd app ports` (updated: `fix(ports): compare bind IPs by canonical value`)

The one open finding from the prior review is now resolved, and no new issues were introduced. **Approving.**

- **[Resolved · was Medium]** `bindIPsOverlap` no longer falls through to a raw string compare for two specific IPs. It now parses both with `net.ParseIP` and compares via `net.IP.Equal`, so equivalent textual forms of the same address are correctly detected as overlapping — e.g. `[2001:db8::1]` vs `[2001:0db8:0:0:0:0:0:1]`, and the IPv4-mapped form `[::ffff:c0a8:0101]` vs `192.168.1.1`. Unparseable bind IPs (e.g. hostnames) still fail safe to non-overlapping, and the wildcard/family handling is unchanged.

### What looks good
- New `TestCollectPortReportConflictIdentityMatrix` cases lock in the fix: equivalent IPv6 literals conflict, IPv4-mapped IPv6 conflicts with the equivalent IPv4 literal, and two distinct concrete IPv6 addresses do not conflict.
- The test assertions were migrated to `testify` `assert`/`require`, matching the repository's stated testing convention.
- The full conflict model — `(BindIP, HostPort, Protocol)` keying, union-find grouping, wildcard/family overlap semantics, `{services, conflicts}` JSON with clean stdout on conflict — remains correct end to end.

All three original findings plus the follow-up Medium are now addressed. Nice work.

_Note: automated build/test were not re-run in place (the local working tree is not checked out at the PR head); this approval is based on direct inspection of the PR-head diff and tests. The change is an 8-line, self-contained correctness fix with matching test coverage._

File: cli/src/cmd/app/commands/ports.go [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+	if isIPv6Wildcard(a) {
+		return bindIPFamily(b) == "ipv6"
+	}
+	if isIPv6Wildcard(b) {
+		return bindIPFamily(a) == "ipv6"
+	}
+
+	ipA := net.ParseIP(a)
+	ipB := net.ParseIP(b)
+	if ipA == nil || ipB == nil {
```
Comment (wbreza) [inline cli/src/cmd/app/commands/ports.go:303]: **[conflict-detection] (Medium)** Hostname bind targets never overlap IP literals, so a real host-port collision is missed. `ParsePortSpec` stores the bind target verbatim (`cli/src/internal/service/port.go:179`) and accepts hostnames — `localhost:3000:8080` yields `BindIP: "localhost"` (`cli/src/internal/service/ports_test.go:83-88`). For two distinct non-empty, non-wildcard targets, `bindIPsOverlap` parses each with `net.ParseIP`; a hostname parses to `nil`, so this line returns `false` (`ports.go:284-286`). As a result `localhost:3000:8080` and `127.0.0.1:3000:8080` on two services are reported as non-conflicting, even though `localhost` resolves to `127.0.0.1` on a standard host — defeating the preflight gate for that case. It under-reports (fail-safe) and true equivalence is resolver/host dependent, hence Medium.

Fix: canonicalize well-known hostnames (at minimum map `localhost` to the loopback address for each family) before the `net.ParseIP` fallthrough, or resolve hostnames and compare effective addresses; keep the current `nil → false` behavior only for genuinely unresolvable names.

File: cli/src/cmd/app/commands/ports.go [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+
+func conflictBindingLabel(c portConflict) string {
+	return fmt.Sprintf("%s/%s", formatHostBinding(c.BindIP, strconv.Itoa(c.HostPort)), c.Protocol)
+}
+
+func formatHostBinding(bindIP, host string) string {
+	if bindIP == "" {
+		return host
+	}
+	if strings.Contains(bindIP, ":") {
```
Comment (wbreza) [inline cli/src/cmd/app/commands/ports.go:416]: **[conflict-label-format] (Low)** Grouped multi-bind conflict labels can render malformed host strings. When a wildcard chains several specific IPs into one conflict group, `summarizeBindIPs` joins them into a comma-separated string (`ports.go:309-323`) that `conflictBindingLabel` passes whole into `formatHostBinding` (`ports.go:372-374`). Because `formatHostBinding` brackets any value containing `:` (`ports.go:380-382`), a summary such as `::, ::1` renders as `[::, ::1]:3000/tcp` in the warning line, and the JSON `portConflict.BindIP` becomes a delimited list rather than a single address — not machine-parseable.

Fix: keep the structured bind IPs on `portConflict` and format each independently, or skip the bracket logic when the value is a summary (contains `,`).

**Review by wbreza — COMMENTED** [review]
## Code Review — Re-review (against current HEAD)

Verified all four prior findings are fixed and their threads resolved. Confirmed against the current merged HEAD; `ports.go` is unchanged by the merge, so the checks below still apply. This round adds two new findings from a full pass over the diff.

### Replied & Resolved by agreement (4)
- Conflicts now keyed on the full `{BindIP, HostPort, Protocol}` identity with wildcard-family overlap handling.
- Bind IP preserved in both text output (via `formatHostBinding`, IPv6 bracketed) and JSON (`bindIP,omitempty`).
- Top-level `conflicts` array plus `cmd.SilenceErrors` in JSON mode keeps stdout valid JSON.
- Canonical IP comparison via `net.ParseIP` + `net.IP.Equal`, including IPv4-mapped IPv6.

All four are correct and covered by tests (the 10-row conflict identity matrix, JSON-shape, and text-render tests).

### New findings this round
| Priority | Count |
|----------|-------|
| Medium   | 1     |
| Low      | 1     |
| **Total**| **2** |

- Medium — Hostname bind targets never overlap IP literals, so a `localhost` vs `127.0.0.1` collision escapes the conflict gate (inline at `ports.go:284`).
- Low — Grouped multi-bind conflict labels can render malformed host strings and a non-parseable JSON `bindIP` (inline at `ports.go:380`).

### What looks good
- Composite conflict identity with union-find clustering is correct and thoroughly tested.
- Wildcard/family overlap semantics (`0.0.0.0` / `::` / empty) match the documented behavior.
- JSON output is stable — empty conflicts marshal as `[]`, not `null`.
- Command cleanly reuses `GetPortMappings` / `ParsePortSpec` so specs resolve the same way `run` sees them.

**Overall Assessment:** Comment — the Medium is a fail-safe under-report limited to hostname bind targets; the command is otherwise solid.

**Review by wbreza — COMMENTED** [review]
## Code Review — Re-review (merge sync `37f40a1e`)

The branch merged `main` again (`ad1d7878` → `37f40a1e`), pulling in unrelated features (remove, hooks, env, logs). The PR's own diff against base is still limited to the six ports files, and `ports.go` / `ports_test.go` are byte-identical to the previously reviewed revision — this merge introduced no substantive change to the command under review.

### Status of prior findings
- All four original findings remain fixed and resolved.
- Two findings from the last round are still open on their existing inline threads and are unchanged by this merge:
  - Medium — Hostname bind targets never overlap IP literals (`ports.go:284`), so a `localhost` vs `127.0.0.1` collision escapes the conflict gate.
  - Low — Grouped multi-bind conflict labels can render malformed host strings and a non-parseable JSON `bindIP` (`ports.go:380`).

No new inline comments are added this round to avoid duplicating those open threads.

### Summary
| Priority | Count (open) |
|----------|--------------|
| Medium   | 1            |
| Low      | 1            |
| **Total**| **2**        |

**Overall Assessment:** Comment — one Medium (fail-safe under-report for hostname bind targets) remains open from the prior round; the merge itself is clean and changes nothing in the ports command.

**Review by wbreza — APPROVED** [review]
## Code Review — Re-review (`134f7fb3`) — Approved

Commit `134f7fb3` (`fix(ports): detect localhost conflicts and render grouped bind labels`) resolves both findings from the prior round. Verified against the diff: `go build ./...` clean and the ports test suite passes.

### Both prior findings fixed
- **Medium — localhost conflict detection** (`ports.go`): `bindIPsOverlap` now short-circuits on a loopback alias via `isLoopbackAlias` + `coversLoopback`, so `localhost` overlaps `127.0.0.1`, `::1`, either wildcard, and the empty all-interfaces bind, while correctly staying separate from routable addresses (`192.168.1.5`), unresolvable hostnames (`db.internal`), and non-canonical loopbacks (`127.0.0.2`). Case-insensitive and order-independent. Covered by `TestBindIPsOverlapHandlesLocalhost` (11 cases).
- **Low — grouped conflict label rendering** (`ports.go`): `portConflict` now keeps bind targets structured as `BindIPs []string` (JSON `bindIPs`), and `conflictBindingLabel` renders each independently, so a wildcard chained to a specific address reads as `[::]:3000/tcp, [::1]:3000/tcp` instead of a bracketed malformed string. All-interfaces groups omit the address. Covered by `TestConflictBindingLabelRendersEachBindIndependently`.

### What looks good
- The loopback branch is scoped to trigger only when a side is literally `localhost`, leaving all other overlap paths unchanged — no regression to the wildcard/family or `net.IP.Equal` logic.
- `coversLoopback` deliberately excludes non-canonical loopbacks and resolver-dependent hostnames rather than guessing — a sound fail-safe choice, clearly documented.
- Test additions map one-to-one to the reported edge cases.

**Overall Assessment:** Approve — all findings from prior rounds are resolved and no new issues were found.

File: cli/src/cmd/app/commands/ports.go (hunk 1 above)
Comment (wbreza) [reply cli/src/cmd/app/commands/ports.go]: Fixed in `134f7fb3`. `bindIPsOverlap` now treats `localhost` as a loopback alias overlapping the loopback literals and wildcards, verified by `TestBindIPsOverlapHandlesLocalhost`. Resolving.

File: cli/src/cmd/app/commands/ports.go (hunk 2 above)
Comment (wbreza) [reply cli/src/cmd/app/commands/ports.go:46]: Fixed in `134f7fb3`. `portConflict` now stores `BindIPs []string` and `conflictBindingLabel` renders each target independently, verified by `TestConflictBindingLabelRendersEachBindIndependently`. Resolving.

File: cli/src/cmd/app/commands/ports.go (hunk 5 above)
Comment (wbreza) [reply cli/src/cmd/app/commands/ports.go:303]: Fixed in `134f7fb3`. `bindIPsOverlap` now treats `localhost` as a loopback alias (`isLoopbackAlias` + `coversLoopback`) overlapping the loopback literals and wildcards, while staying separate from routable/non-canonical/unresolvable targets. Verified by `TestBindIPsOverlapHandlesLocalhost`. Resolved.

File: cli/src/cmd/app/commands/ports.go (hunk 6 above)
Comment (wbreza) [reply cli/src/cmd/app/commands/ports.go:416]: Fixed in `134f7fb3`. `portConflict` now stores `BindIPs []string` and `conflictBindingLabel` renders each target independently (e.g. `[::]:3000/tcp, [::1]:3000/tcp`). Verified by `TestConflictBindingLabelRendersEachBindIndependently`. Resolved.

File: cli/src/cmd/app/commands/ports.go (hunk 5 above)
Comment (jongio) [reply cli/src/cmd/app/commands/ports.go:303]: Confirmed and fixed in 134f7fb3. ParsePortSpec stores the bind target verbatim at port.go:179, so `localhost` reached `bindIPsOverlap`, `net.ParseIP` returned nil, and it fell through to `return false`. `localhost:3000:8080` and `127.0.0.1:3000:8080` were reported as non-conflicting.

`localhost` is now compared as a loopback alias: it overlaps the canonical loopback literals `127.0.0.1` and `::1` plus both wildcards. I scoped it to the canonical addresses rather than all of `127.0.0.0/8`, since `localhost` doesn't resolve to `127.0.0.2` and flagging that would be a false positive. This also fixed a related under-report: `::` vs `localhost` previously returned false because `bindIPFamily` reports `unknown` for a hostname. Other hostnames keep the `nil` to `false` behavior you suggested. Covered by `TestBindIPsOverlapHandlesLocalhost`, 11 cases.

File: cli/src/cmd/app/commands/ports.go (hunk 6 above)
Comment (jongio) [reply cli/src/cmd/app/commands/ports.go:416]: Confirmed and fixed in 134f7fb3. Took the first option you offered: `portConflict` keeps the addresses structured as `bindIPs []string` instead of flattening them through `summarizeBindIPs`, and `conflictBindingLabel` formats each one independently.

`[::, ::1]:3000/tcp` now renders as `[::]:3000/tcp, [::1]:3000/tcp`, and the JSON field is a real array rather than a delimited string. A group that only binds all interfaces omits the field, matching `portBinding.BindIP`. `summarizeBindIPs` is deleted. Covered by `TestConflictBindingLabelRendersEachBindIndependently`.
